ios/uPortMobile/Classes/EthereumSigner.m

```diff
@@ -0,0 +1,203 @@
+//
+// uPort Mobile App is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
+
+#import "EthereumSigner.h"
+
+#import "CoreBitcoin/CoreBitcoin+Categories.h"
+#import "CoreBitcoin/BTCKey.h"
+#include <openssl/bn.h>
+#include <openssl/ec.h>
+#include <openssl/obj_mac.h>
+
+NSMutableData *compressedPublicKey(EC_KEY *key) {
+  if (!key) return nil;
+  EC_KEY_set_conv_form(key, POINT_CONVERSION_COMPRESSED);
+  int length = i2o_ECPublicKey(key, NULL);
+  if (!length) return nil;
+  NSCAssert(length <= 65, @"Pubkey length must be up to 65 bytes.");
+  NSMutableData* data = [[NSMutableData alloc] initWithLength:length];
+  unsigned char* bytes = [data mutableBytes];
+  if (i2o_ECPublicKey(key, &bytes) != length) return nil;
+  return data;
+}
+
+
+NSDictionary *ethereumSignature(BTCKey *keypair, NSData *hash, NSData *chainId) {
+  NSDictionary *sig = genericSignature(keypair, hash, YES);
+  if (!sig) return NULL;
+  NSData *rData = (NSData *)sig[@"r"];
+  NSData *sData = (NSData *)sig[@"s"];
+  BN_ULONG base = 0x1b; // pre-EIP155
+  if (chainId) {
+    BIGNUM *v = BN_new(); BN_bin2bn(chainId.bytes, chainId.length, v);
+    // TODO support longer chainIDs
+    base = BN_get_word(v) * 2 + 35;
+    BN_clear_free(v);
+  }
+  
+  NSDictionary *signatureDictionary = @{ @"v": @(base + [sig[@"recoveryParam"] intValue]),
+                                         @"r": [rData base64EncodedStringWithOptions:0],
+                                         @"s":[sData base64EncodedStringWithOptions:0]};
+  return signatureDictionary;
+}
+
+
+NSDictionary *genericSignature(BTCKey *keypair, NSData *hash, BOOL lowS) {
+    BTCBigNumber* privkeyBN = [[BTCBigNumber alloc] initWithUnsignedBigEndian:[keypair privateKey]];
+    BTCBigNumber* n = [BTCCurvePoint curveOrder];
+  
+    NSMutableData* kdata = [keypair signatureNonceForHash:hash];
+    BTCMutableBigNumber* k = [[BTCMutableBigNumber alloc] initWithUnsignedBigEndian:kdata];
+    [k mod:n]; // make sure k belongs to [0, n - 1]
+    
+    if (BN_cmp(k.BIGNUM, [BTCBigNumber zero].BIGNUM) <= 0 && BN_cmp(k.BIGNUM, n.BIGNUM) >= 0) return NULL;
+
+    BTCDataClear(kdata);
+  
+    BTCCurvePoint* K = [[BTCCurvePoint generator] multiply:k];
+    // K was infinity and does not work, redo it
+    if ([K isInfinity])  return NULL;
+    BTCMutableBigNumber* Kx = [K.x mutableCopy];
+    [Kx mod: n];
+    
+    BTCBigNumber* hashBN = [[BTCBigNumber alloc] initWithUnsignedBigEndian:hash];
+  
+    // Compute s = (k^-1)*(h + Kx*privkey)
+  
+    BTCMutableBigNumber* signatureBN = [[[[privkeyBN mutableCopy] multiply:Kx mod:n] add:hashBN mod: n] multiply:[k inverseMod:n] mod:n];
```

Comment:
Remove space between `:` and `n` => `[[[[privkeyBN mutableCopy] multiply:Kx mod:n] add:hashBN mod:n] multiply:[k inverseMod:n] mod:n];`